### PR TITLE
Ed Services Report

### PR DESCRIPTION
**Pull request by johnalbaker**
In support of the ed services report this PR imports the coursemode_coursemode table into HIVE.

### CODE REVIEW

**Comment by johnalbaker** [comment]
PR includes features supporting the creation of the ed services report :
- Importing course_mode_coursemode table for mode specific information
- Importing student_course_enrollment to enrollment information
- Building ed services report csv and uploading to HIVE / S3

@mulby @brianhw

File: edx/analytics/tasks/__init__.py [hunk 1]
```diff
@@ -1 +1 @@
-EXTENSION_NAMESPACE = "edx.analytics.tasks"
+EXTENSION_NAMESPACE = "edx.analytics.tasks"
```
Comment (mulby) [inline edx/analytics/tasks/__init__.py:1]: nit: you can revert this change

File: edx/analytics/tasks/database_imports.py [hunk 2]
```diff
@@ -991,6 +991,32 @@ def columns(self):
         ]
 
 
+class ImportCourseInformation(ImportMysqlToHiveTableTask):
```
Comment (mulby) [inline edx/analytics/tasks/database_imports.py]: I would prefer this was named in a way that more closely matched the table being imported. There is a lot of "course information" out there, which makes it unclear which bit this is actually importing.

File: edx/analytics/tasks/reports/data_import/coursemode_imports.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import luigi
+import luigi.hdfs
+
+from edx.analytics.tasks.util.overwrite import OverwriteOutputMixin
+from edx.analytics.tasks.database_imports import (DatabaseImportMixin,ImportCourseInformation)
+
+class CourseModeTableTask(DatabaseImportMixin, OverwriteOutputMixin, luigi.WrapperTask):
```
Comment (mulby) [inline edx/analytics/tasks/reports/data_import/coursemode_imports.py:7]: I don't understand why this task is necessary... it seems like any task that depends on this task could instead just depend on the one task that it wraps, removing the layer of indirection.

File: edx/analytics/tasks/reports/financial_report/ed_services_financial_report.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    Builds the Course and Enrollment data to satisfy the Ed Services report.
+    """
+    def requires(self):
+        kwargs = {
+            'num_mappers': self.num_mappers,
+            'verbose': self.verbose,
+            'import_date': self.import_date,
+            'overwrite': self.overwrite,
+        }
+        yield (
```
Comment (mulby) [inline edx/analytics/tasks/reports/financial_report/ed_services_financial_report.py:22]: I think we can just move all of these requirements directly into the task below, no need for a wrapper task here.

File: edx/analytics/tasks/reports/financial_report/ed_services_financial_report.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+    def output(self):
+        return [task.output() for task in self.requires()]
+
+
+class BuildEdServicesReportTask(DatabaseImportMixin, HiveTableFromQueryTask):
+    """
+    Builds the financial report delivered to Ed Services.
+
+    """
+    destination = luigi.Parameter(
```
Comment (mulby) [inline edx/analytics/tasks/reports/financial_report/ed_services_financial_report.py]: These parameters are not necessary since they are defined in DatabaseImportMixin

File: edx/analytics/tasks/reports/financial_report/ed_services_financial_report.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+
+    @property
+    def columns(self):
+        return [
+            ('course_id', 'STRING'),
+            ('mode_slug', 'STRING'),
+            ('suggested_prices', 'STRING'),
+            ('expiration_datetime', 'TIMESTAMP'),
+            ('total_currently_enrolled', 'INT'),
+            ('Audit_currently_enrolled', 'INT'),
```
Comment (mulby) [inline edx/analytics/tasks/reports/financial_report/ed_services_financial_report.py]: nit: I prefer the column casing is consistent, ideally following the convention of all other column names used in the warehouse (lower_case_with_underscores).

File: edx/analytics/tasks/reports/financial_report/ed_services_financial_report.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    def partition(self):
+        return HivePartition('dt', self.import_date.isoformat())  # pylint: disable=no-member
+
+    @property
+    def insert_query(self):
+        return """
+            SELECT
+                VP_COURSES.course_id,
+                VP_COURSES.mode_slug, -- first one of the modes, if any are set up
+                VP_COURSES.suggested_prices,
```
Comment (mulby) [inline edx/analytics/tasks/reports/financial_report/ed_services_financial_report.py:103]: We should disambiguate the requirements here, but I think the intent was to put min_price here instead of the suggested prices.

File: edx/analytics/tasks/reports/financial_report/ed_services_financial_report.py [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+                VP_COURSES.expiration_datetime,
+                ALL_ENROLLS.total_currently_enrolled,
+                ALL_ENROLLS.Audit_currently_enrolled,
+                ALL_ENROLLS.Honor_currently_enrolled,
+                ALL_ENROLLS.Verified_currently_enrolled,
+                ALL_ENROLLS.Professional_currently_enrolled,
+                ALL_ENROLLS.No_id_professional_currently_enrolled,
+                ALL_ENROLLS.ERROR_currently_enrolled, -- this should be 0 everywhere, unless we're missing a mode
+                COALESCE(seats.refunded_seats,0) refunded_seat_count,
```
Comment (mulby) [inline edx/analytics/tasks/reports/financial_report/ed_services_financial_report.py:112]: Since these columns are defined in the subquery I would prefer to just change their names in the subquery instead of aliasing them at this level. I think it would make the query easier to grok.

File: edx/analytics/tasks/reports/financial_report/ed_services_financial_report.py [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+                COALESCE(seats.net_amount,0) net_seat_revenue,
+                COALESCE(seats.net_seats,0) net_seat_count,
+                COALESCE(donations.donations,0) donation_count,
+                COALESCE(donations.net_donation_revenue,0) net_donation_revenue
+
+            FROM
+            (
+                -- Course Information --
+                select
+                    cmc.*
```
Comment (mulby) [inline edx/analytics/tasks/reports/financial_report/ed_services_financial_report.py:123]: I would prefer we explicitly list the columns here since it is more tolerant of schema evolution.

**Comment by mulby** [comment]
I think we can merge this into the branch as is :+1: 

We can address all of these comments when we go to productize the spike.

File: edx/analytics/tasks/database_imports.py (hunk 2 above)
Comment (johnalbaker) [inline edx/analytics/tasks/database_imports.py]: I changed this to ImportCourseModeTask.

File: edx/analytics/tasks/reports/financial_report/ed_services_financial_report.py (hunk 6 above)
Comment (johnalbaker) [inline edx/analytics/tasks/reports/financial_report/ed_services_financial_report.py]: Changed them all to lower case.

File: edx/analytics/tasks/reports/financial_report/ed_services_financial_report.py (hunk 5 above)
Comment (johnalbaker) [inline edx/analytics/tasks/reports/financial_report/ed_services_financial_report.py]: Removed.

File: edx/analytics/tasks/reports/data_import/coursemode_imports.py (hunk 3 above)
Comment (johnalbaker) [inline edx/analytics/tasks/reports/data_import/coursemode_imports.py:7]: Changed: I pointed directly to ImportCourseModeTask in database_imports.
